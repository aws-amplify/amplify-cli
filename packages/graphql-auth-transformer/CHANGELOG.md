# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [3.9.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.7.5...graphql-auth-transformer@3.9.0) (2019-08-02)


### Features

* adding amplify cli predictions category ([#1936](https://github.com/aws-amplify/amplify-cli/issues/1936)) ([b7b7c2c](https://github.com/aws-amplify/amplify-cli/commit/b7b7c2c))





# [3.8.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.7.5...graphql-auth-transformer@3.8.0) (2019-07-31)


### Features

* adding amplify cli predictions category ([#1936](https://github.com/aws-amplify/amplify-cli/issues/1936)) ([b7b7c2c](https://github.com/aws-amplify/amplify-cli/commit/b7b7c2c))





## [3.7.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.7.4...graphql-auth-transformer@3.7.5) (2019-07-24)

**Note:** Version bump only for package graphql-auth-transformer





## [3.7.4](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.7.2...graphql-auth-transformer@3.7.4) (2019-06-30)

**Note:** Version bump only for package graphql-auth-transformer





## [3.7.2](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.7.1...graphql-auth-transformer@3.7.2) (2019-06-26)

**Note:** Version bump only for package graphql-auth-transformer





## [3.7.1](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.7.0...graphql-auth-transformer@3.7.1) (2019-06-12)

**Note:** Version bump only for package graphql-auth-transformer





# [3.7.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.6.4...graphql-auth-transformer@3.7.0) (2019-05-29)


### Features

* feature/[@key](https://github.com/key) ([#1463](https://github.com/aws-amplify/amplify-cli/issues/1463)) ([00ed819](https://github.com/aws-amplify/amplify-cli/commit/00ed819))





## [3.6.4](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.6.3...graphql-auth-transformer@3.6.4) (2019-05-21)

**Note:** Version bump only for package graphql-auth-transformer





## [3.6.3](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.6.2...graphql-auth-transformer@3.6.3) (2019-05-17)

**Note:** Version bump only for package graphql-auth-transformer





## [3.6.2](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.6.0...graphql-auth-transformer@3.6.2) (2019-05-07)

**Note:** Version bump only for package graphql-auth-transformer





## [3.6.1](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.6.0...graphql-auth-transformer@3.6.1) (2019-05-06)

**Note:** Version bump only for package graphql-auth-transformer





# [3.6.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.5.1...graphql-auth-transformer@3.6.0) (2019-04-16)


### Features

* **field-level-auth:** Add field level auth support via the [@auth](https://github.com/auth) directive ([#1262](https://github.com/aws-amplify/amplify-cli/issues/1262)) ([3b1c600](https://github.com/aws-amplify/amplify-cli/commit/3b1c600)), closes [#1043](https://github.com/aws-amplify/amplify-cli/issues/1043)





## [3.5.1](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.4.0...graphql-auth-transformer@3.5.1) (2019-04-09)

**Note:** Version bump only for package graphql-auth-transformer





# [3.4.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.9...graphql-auth-transformer@3.4.0) (2019-04-03)


### Bug Fixes

* **graphql-auth-transformer:** conditional group expression ([#1186](https://github.com/aws-amplify/amplify-cli/issues/1186)) ([83ef244](https://github.com/aws-amplify/amplify-cli/commit/83ef244)), closes [#360](https://github.com/aws-amplify/amplify-cli/issues/360)


### Features

* support for provisioning Cognito Hosted UI and support CRUD operations in Storage and API categories ([729b0de](https://github.com/aws-amplify/amplify-cli/commit/729b0de))





## [3.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.8...graphql-auth-transformer@3.0.9) (2019-03-22)

**Note:** Version bump only for package graphql-auth-transformer





## [3.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.7...graphql-auth-transformer@3.0.8) (2019-03-05)

**Note:** Version bump only for package graphql-auth-transformer





## [3.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.6...graphql-auth-transformer@3.0.7) (2019-02-20)

**Note:** Version bump only for package graphql-auth-transformer





## [3.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.5...graphql-auth-transformer@3.0.6) (2019-02-12)

**Note:** Version bump only for package graphql-auth-transformer





## [3.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.3-beta.0...graphql-auth-transformer@3.0.5) (2019-02-11)

**Note:** Version bump only for package graphql-auth-transformer





## [3.0.3](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.3-beta.0...graphql-auth-transformer@3.0.3) (2019-02-11)

**Note:** Version bump only for package graphql-auth-transformer





## [3.0.3-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@3.0.2...graphql-auth-transformer@3.0.3-beta.0) (2019-02-11)

**Note:** Version bump only for package graphql-auth-transformer





<a name="2.0.0-multienv.2"></a>
# [2.0.0-multienv.2](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.34-multienv.1...graphql-auth-transformer@2.0.0-multienv.2) (2018-12-31)


### Bug Fixes

* update grahql transformer package versions for multienv ([8b4b2bd](https://github.com/aws-amplify/amplify-cli/commit/8b4b2bd))




<a name="1.0.34-multienv.1"></a>
## [1.0.34-multienv.1](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.34-multienv.0...graphql-auth-transformer@1.0.34-multienv.1) (2018-12-19)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.34-multienv.0"></a>
## [1.0.34-multienv.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.33...graphql-auth-transformer@1.0.34-multienv.0) (2018-11-16)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.33"></a>
## [1.0.33](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.33-beta.0...graphql-auth-transformer@1.0.33) (2018-11-09)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.33-beta.0"></a>
## [1.0.33-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.12...graphql-auth-transformer@1.0.33-beta.0) (2018-11-09)


### Bug Fixes

* **graphql-auth-transformer:** Modifies the name of the user pool resource ([d613481](https://github.com/aws-amplify/amplify-cli/commit/d613481))




<a name="1.0.32"></a>
## [1.0.32](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.32-beta.0...graphql-auth-transformer@1.0.32) (2018-11-05)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.32-beta.0"></a>
## [1.0.32-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.12...graphql-auth-transformer@1.0.32-beta.0) (2018-11-05)


### Bug Fixes

* **graphql-auth-transformer:** Modifies the name of the user pool resource ([d613481](https://github.com/aws-amplify/amplify-cli/commit/d613481))




<a name="1.0.31"></a>
## [1.0.31](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.12...graphql-auth-transformer@1.0.31) (2018-11-02)


### Bug Fixes

* **graphql-auth-transformer:** Modifies the name of the user pool resource ([d613481](https://github.com/aws-amplify/amplify-cli/commit/d613481))




<a name="1.0.30"></a>
## [1.0.30](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.30-beta.0...graphql-auth-transformer@1.0.30) (2018-11-02)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.30-beta.0"></a>
## [1.0.30-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.12...graphql-auth-transformer@1.0.30-beta.0) (2018-11-02)


### Bug Fixes

* **graphql-auth-transformer:** Modifies the name of the user pool resource ([d613481](https://github.com/aws-amplify/amplify-cli/commit/d613481))




<a name="1.0.29"></a>
## [1.0.29](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.29-beta.0...graphql-auth-transformer@1.0.29) (2018-10-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.29-beta.0"></a>
## [1.0.29-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.12...graphql-auth-transformer@1.0.29-beta.0) (2018-10-23)


### Bug Fixes

* **graphql-auth-transformer:** Modifies the name of the user pool resource ([d613481](https://github.com/aws-amplify/amplify-cli/commit/d613481))




<a name="1.0.28"></a>
## [1.0.28](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.28-beta.0...graphql-auth-transformer@1.0.28) (2018-10-18)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.28-beta.0"></a>
## [1.0.28-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.12...graphql-auth-transformer@1.0.28-beta.0) (2018-10-12)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.12"></a>
## [1.0.12](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.11...graphql-auth-transformer@1.0.12) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.11"></a>
## [1.0.11](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.10...graphql-auth-transformer@1.0.11) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.10"></a>
## [1.0.10](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.9...graphql-auth-transformer@1.0.10) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.9"></a>
## [1.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.8...graphql-auth-transformer@1.0.9) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.8"></a>
## [1.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.7...graphql-auth-transformer@1.0.8) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.7"></a>
## [1.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.6...graphql-auth-transformer@1.0.7) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.6"></a>
## [1.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.5...graphql-auth-transformer@1.0.6) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.5"></a>
## [1.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-auth-transformer@1.0.4...graphql-auth-transformer@1.0.5) (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.4"></a>
## 1.0.4 (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.3"></a>
## 1.0.3 (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.2"></a>
## 1.0.2 (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer

<a name="1.0.1"></a>
## 1.0.1 (2018-08-23)




**Note:** Version bump only for package graphql-auth-transformer
