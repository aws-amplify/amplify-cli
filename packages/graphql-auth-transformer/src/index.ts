import { ModelAuthTransformer } from './ModelAuthTransformer';
export * from './ModelAuthTransformer';
export default ModelAuthTransformer;
