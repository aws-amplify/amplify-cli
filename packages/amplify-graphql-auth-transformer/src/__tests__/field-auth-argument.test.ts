import { AuthTransformer } from '../graphql-auth-transformer';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import { ResourceConstants } from 'graphql-transformer-common';
import { AppSyncAuthConfiguration } from '@aws-amplify/graphql-transformer-interfaces';

test('subscriptions are only generated if the respective mutation operation exists', () => {
  const validSchema = `
      type Salary
        @model
        @auth(rules: [
                {allow: owner},
                {allow: groups, groups: ["Moderator"]}
            ]) {
        id: ID!
        wage: Int
        owner: String
        secret: String @auth(rules: [{allow: owner}])
      }`;
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [],
  };
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  // expect to generate subscription resolvers for create and update only
  expect(out).toBeDefined();
  expect(out.rootStack.Resources[ResourceConstants.RESOURCES.GraphQLAPILogicalID].Properties.AuthenticationType).toEqual(
    'AMAZON_COGNITO_USER_POOLS',
  );
  expect(out.resolvers['Salary.secret.res.vtl']).toContain('#if( $operation == "Mutation" )');

  expect(out.resolvers['Mutation.createSalary.res.vtl']).toContain('$util.qr($ctx.result.put("__operation", "Mutation"))');
  expect(out.resolvers['Mutation.updateSalary.res.vtl']).toContain('$util.qr($ctx.result.put("__operation", "Mutation"))');
  expect(out.resolvers['Mutation.deleteSalary.res.vtl']).toContain('$util.qr($ctx.result.put("__operation", "Mutation"))');
});

test('per-field @auth without @model', () => {
  const validSchema = `
    type Query {
      listContext: String @auth(rules: [{ allow: groups, groups: ["Allowed"] }, { allow: private, provider: iam }])
    }`;
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{ authenticationType: 'AWS_IAM' }],
  };
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();

  const resources = out.rootStack.Resources;
  const authPolicyIdx = Object.keys(out.rootStack.Resources).find(r => r.includes('AuthRolePolicy'));
  expect(resources[authPolicyIdx]).toMatchSnapshot();
  expect(out.resolvers['Query.listContext.req.vtl']).toContain(
    '#set( $staticGroupRoles = [{"claim":"cognito:groups","entity":"Allowed"}] )',
  );
});

test('error on non null fields which need resolvers', () => {
  const invalidSchema = `
    type Post @model @auth(rules: [{ allow: groups, groups: ["admin"] }]) {
      id: ID!
      name: String!
      ssn: String! @auth(rules: [{ allow: owner }])
    }
  `;
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{ authenticationType: 'AWS_IAM' }],
  };
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  expect(() => transformer.transform(invalidSchema)).toThrowErrorMatchingSnapshot();
});

test('does not generate field resolvers when private rule takes precedence over provider-related rules', () => {
  const validSchema = `
  type Student @model @auth(rules: [{ allow: private, provider: userPools }, { allow: private, provider: iam }]) {
    id: ID!
    name: String!
    ssn: String @auth(rules: [{ allow: owner }])
  }`;

  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{ authenticationType: 'AWS_IAM' }],
  };
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.resolvers['Student.ssn.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Student.ssn.res.vtl']).toMatchSnapshot();
  for (let field of ['id', 'name']) {
    expect(out.resolvers[`Student.${field}.req.vtl`]).toBeUndefined();
    expect(out.resolvers[`Student.${field}.res.vtl`]).toBeUndefined();
  }
});

test('generates field resolver for other provider rules even if private removes all provided-related rules', () => {
  const validSchema = `
  type Student @model @auth(rules: [{ allow: private, provider: userPools }]) {
    id: ID
    name: String
    ssn: String @auth(rules: [{ allow: owner }, { allow: private, provider: iam }])
  }`;

  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{ authenticationType: 'AWS_IAM' }],
  };
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();
  expect(out.resolvers['Student.ssn.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Student.ssn.res.vtl']).toMatchSnapshot();
  for (let field of ['id', 'name']) {
    expect(out.resolvers[`Student.${field}.req.vtl`]).toBeDefined();
    expect(out.resolvers[`Student.${field}.res.vtl`]).toBeDefined();
  }
});

describe('subscription disabled and userPools configured', () => {
  describe('with non-nullable (required) fields', () => {
    describe('top level private and field level owner auth', () => {
      test('generates field resolver for required field with expected owner claim', () => {
        const validSchema = `
          type Student
            @model(subscriptions: { level: off })
            @auth(rules: [
              { allow: private }
            ])
          {
            id: ID
            name: String
            ssn: String! @auth(rules: [{ allow: owner }])
          }`;

        const authConfig: AppSyncAuthConfiguration = {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        };
        const transformer = new GraphQLTransform({
          authConfig,
          transformers: [new ModelTransformer(), new AuthTransformer()],
        });
        const out = transformer.transform(validSchema);
        expect(out).toBeDefined();

        expect(out.resolvers['Student.ssn.req.vtl']).toMatchSnapshot();
        expect(out.resolvers['Student.ssn.req.vtl']).toContain(`#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get("username"), $util.defaultIfNull($ctx.identity.claims.get("cognito:username"), "___xamznone____")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end`);
      });
    });

    describe('top level private and field level group auth', () => {
      test('generates field resolver for required field with expected group role', () => {
        const validSchema = `
          type Student
            @model(subscriptions: { level: off })
            @auth(rules: [
              { allow: private }
            ])
          {
            id: ID
            name: String
            ssn: String! @auth(rules: [{ allow: groups, groups: ["admin"] }])
          }`;

        const authConfig: AppSyncAuthConfiguration = {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        };
        const transformer = new GraphQLTransform({
          authConfig,
          transformers: [new ModelTransformer(), new AuthTransformer()],
        });
        const out = transformer.transform(validSchema);
        expect(out).toBeDefined();

        expect(out.resolvers['Student.ssn.req.vtl']).toMatchSnapshot();
        expect(out.resolvers['Student.ssn.req.vtl']).toContain(`#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{"claim":"cognito:groups","entity":"admin"}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #set( $isAuthorized = true )
        #break
      #end
    #end
  #end
#end`);
      });
    });
  });

  describe('with nullable fields', () => {
    describe('top level private and field level owner auth', () => {
      test('generates field resolver for field with expected owner claim', () => {
        const validSchema = `
          type Student
            @model(subscriptions: { level: off })
            @auth(rules: [
              { allow: private }
            ])
          {
            id: ID
            name: String
            ssn: String @auth(rules: [{ allow: owner }])
          }`;

        const authConfig: AppSyncAuthConfiguration = {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        };
        const transformer = new GraphQLTransform({
          authConfig,
          transformers: [new ModelTransformer(), new AuthTransformer()],
        });
        const out = transformer.transform(validSchema);
        expect(out).toBeDefined();

        expect(out.resolvers['Student.ssn.req.vtl']).toMatchSnapshot();
        expect(out.resolvers['Student.ssn.req.vtl']).toContain(`#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get("username"), $util.defaultIfNull($ctx.identity.claims.get("cognito:username"), "___xamznone____")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end`);
      });
    });

    describe('top level private and field level group auth', () => {
      test('generates field resolver for field with expected group roles', () => {
        const validSchema = `
          type Student
            @model(subscriptions: { level: off })
            @auth(rules: [
              { allow: private }
            ])
          {
            id: ID
            name: String
            ssn: String @auth(rules: [{ allow: groups, groups: ["admin"] }])
          }`;

        const authConfig: AppSyncAuthConfiguration = {
          defaultAuthentication: {
            authenticationType: 'AMAZON_COGNITO_USER_POOLS',
          },
          additionalAuthenticationProviders: [],
        };
        const transformer = new GraphQLTransform({
          authConfig,
          transformers: [new ModelTransformer(), new AuthTransformer()],
        });
        const out = transformer.transform(validSchema);
        expect(out).toBeDefined();

        expect(out.resolvers['Student.ssn.req.vtl']).toMatchSnapshot();
        expect(out.resolvers['Student.ssn.req.vtl']).toContain(`#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{"claim":"cognito:groups","entity":"admin"}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #set( $isAuthorized = true )
        #break
      #end
    #end
  #end
#end`);
      });
    });
  });
});
