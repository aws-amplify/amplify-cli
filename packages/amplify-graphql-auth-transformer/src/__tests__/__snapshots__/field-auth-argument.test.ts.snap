// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not generate field resolvers when private rule takes precedence over provider-related rules 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"IAM Authorization\\" )
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;

exports[`does not generate field resolvers when private rule takes precedence over provider-related rules 2`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source.ssn)
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`error on non null fields which need resolvers 1`] = `"Because \\"Post\\" has a field-level authorization rule and subscriptions are enabled, you need to either apply field-level authorization rules to all required fields where all rules have read access [\\"id\\",\\"name\\",\\"ssn\\"], make those fields nullable, or disable subscriptions for \\"Post\\" (setting level to off or public)."`;

exports[`generates field resolver for other provider rules even if private removes all provided-related rules 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"IAM Authorization\\" )
  #if( !$isAuthorized )
    #if( $ctx.identity.userArn == $ctx.stash.authRole )
      #set( $isAuthorized = true )
    #end
  #end
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;

exports[`generates field resolver for other provider rules even if private removes all provided-related rules 2`] = `
"## [Start] Checking for allowed operations which can return this field. **
#set( $operation = $util.defaultIfNull($ctx.source.get(\\"__operation\\"), null) )
#if( $operation == \\"Mutation\\" )
  $util.toJson(null)
#else
  $util.toJson($context.source.ssn)
#end
## [End] Checking for allowed operations which can return this field. **"
`;

exports[`per-field @auth without @model 1`] = `
Object {
  "Properties": Object {
    "Description": "",
    "Path": "/",
    "PolicyDocument": Object {
      "Statement": Array [
        Object {
          "Action": "appsync:GraphQL",
          "Effect": "Allow",
          "Resource": Object {
            "Fn::Sub": Array [
              "arn:aws:appsync:\${AWS::Region}:\${AWS::AccountId}:apis/\${apiId}/types/\${typeName}/fields/\${fieldName}",
              Object {
                "apiId": Object {
                  "Fn::GetAtt": Array [
                    "GraphQLAPI",
                    "ApiId",
                  ],
                },
                "fieldName": "listContext",
                "typeName": "Query",
              },
            ],
          },
        },
      ],
      "Version": "2012-10-17",
    },
    "Roles": Array [
      Object {
        "Ref": "authRoleName",
      },
    ],
  },
  "Type": "AWS::IAM::ManagedPolicy",
}
`;

exports[`subscription disabled and userPools configured with non-nullable (required) fields top level private and field level group auth generates field resolver for required field with expected group role 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{\\"claim\\":\\"cognito:groups\\",\\"entity\\":\\"admin\\"}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #set( $isAuthorized = true )
        #break
      #end
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;

exports[`subscription disabled and userPools configured with non-nullable (required) fields top level private and field level owner auth generates field resolver for required field with expected owner claim 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;

exports[`subscription disabled and userPools configured with nullable fields top level private and field level group auth generates field resolver for field with expected group roles 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{\\"claim\\":\\"cognito:groups\\",\\"entity\\":\\"admin\\"}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #set( $isAuthorized = true )
        #break
      #end
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;

exports[`subscription disabled and userPools configured with nullable fields top level private and field level owner auth generates field resolver for field with expected owner claim 1`] = `
"## [Start] Field Authorization Steps. **
#set( $isAuthorized = false )
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.source.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), \\"___xamznone____\\")) )
    #if( $ownerEntity0 == $ownerClaim0 )
      #set( $isAuthorized = true )
    #end
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Field Authorization Steps. **"
`;
